ma_gym/envs/traffic_junction/traffic_junction.py

```diff
@@ -125,25 +271,171 @@ def __create_grid(self):
 
         return _grid
 
+    def step(self, agents_action):
+        """
+        Performs an action in the environment and steps forward. At each step a new agent enters the road by
+        one of the 4 gates according to a probability "_arrive_prob". A "ncoll" reward is given to an agent if it
+        collides and all of them receive "-0.01*step_n" to avoid traffic jams.
+
+        :param agents_action: list of actions of all the agents to perform in the environment
+        :type agents_action: list
+
+        :return: agents observations, rewards, if agents are done and additional info
+        :rtype: tuple
+        """
+        assert len(agents_action) == self.n_agents, \
+            "Invalid action! It was expected to be list of {}" \
+            " dimension but was found to be of {}".format(self.n_agents, len(agents_action))
+
+        # Todo: Add check to validate each sub-action. It should be a valid action.
+        # This must be done before position of any car is updated
+
+        self._step_count += 1
+        rewards = [0 for _ in range(self.n_agents)]  # initialize rewards array
+
+        # checks if there is a collision; this is done in the __update_agent_pos method
+        for agent_i, action in enumerate(agents_action):
+            if not (self._agent_dones[agent_i]) and self._on_the_road[agent_i]:
+                col_reward = self.__update_agent_pos(agent_i, action)
+                rewards[agent_i] += col_reward
+
+        # TODO remove self.curr_cars_count < self._n_max here, might not be needed if we play with the on the road
+        # adds new car according to the probability _arrive_prob
+        if random.uniform(0, 1) < self._arrive_prob and self.curr_cars_count < self._n_max:
+            for agent_i in range(self.n_agents):
+                if self.__is_gate_free() and not self._on_the_road[agent_i]:
```

Comment:
Instead of traversing through n_agents at every time step, we can do the following:

1. Maintain a  list of unassigned cars: Initially, all cars are in the unassigned list (stack) and we simply pop a car , every time it's added to the entry gate. And, every time, it finishes its route, we push it back to the list.
2. Get a List of Free gates and choose a free gate randomly.
3.  Pop a car from the unassigned list and put it on the gate.